src/sync/impl/sync_client.hpp

```diff
@@ -33,11 +35,23 @@ struct SyncClient {
 
     SyncClient(std::unique_ptr<util::Logger> logger,
                ReconnectMode reconnect_mode = ReconnectMode::normal,
-               bool verify_ssl = true)
+               bool verify_ssl = true,
+               BindingCallbackThreadObserver *client_thread_listener = nullptr)
     : client(make_client(*logger, reconnect_mode, verify_ssl)) // Throws
     , m_logger(std::move(logger))
-    , m_thread([this] {
-        client.run();
+    , m_thread([this, client_thread_listener] {
+        if (client_thread_listener)
+            client_thread_listener->did_create_thread();
+        try {
+            client.run();
+        }
+        catch (...) {
+            if (client_thread_listener)
```

Comment:
It'd be cleaner and more idiomatic to use a scope exit handler for this purpose rather than explicit try / catch.

```c++
if (client_thread_listener)
    client_thread_listener->did_create_thread();

auto will_destroy_thread = util::make_scope_exit([&]() noexcept {
    if (client_thread_listener)
        client_thread_listener->will_destroy_thread();
});

client.run(); // Throws
```

The destructor of `will_destroy_thread` will be invoked when the function returns or if the stack is unwound due to an exception, ensuring that the function is called in both cases.